sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -342,6 +384,103 @@ pi_result piDeviceGetInfo(pi_device device, pi_device_info paramName,
     return PI_SUCCESS;
   }
 
+  case PI_DEVICE_INFO_PARTITION_PROPERTIES: {
+    // SYCL spec says: if this SYCL device cannot be partitioned into at least
+    // two sub devices then the returned vector must be empty.
+    pi_uint32 partitionMaxSubDevices = 0;
+    if (device->subLevel == -1)
+      device->subLevel = getSubLevel(device);
+    if (device->isRootDevice()) {
+      clGetDeviceInfo(
+          cast<cl_device_id>(device), CL_DEVICE_PARTITION_MAX_SUB_DEVICES,
+          sizeof(partitionMaxSubDevices), &partitionMaxSubDevices, nullptr);
+    } else if (device->isSubDevice()) {
+      // find out number of CCSes
+      bool supported = false;
+      cl_int ret_err = CL_SUCCESS;
+      ret_err =
+          checkDeviceExtensions(cast<cl_device_id>(device),
+                                {"cl_intel_command_queue_families"}, supported);
+      if (ret_err != CL_SUCCESS)
+        return static_cast<pi_result>(ret_err);
+      if (!supported) {
+        std::cout
+            << "This device does not support cl_intel_command_queue_families"
+            << std::endl;
+        return ReturnValue(pi_device_partition_property{0});
+      }
+      cl_queue_family_properties_intel qfprops[3];
+      size_t qsize = 0;
+      clGetDeviceInfo(
+          cast<cl_device_id>(device), CL_DEVICE_QUEUE_FAMILY_PROPERTIES_INTEL,
+          3*sizeof(cl_queue_family_properties_intel), qfprops, &qsize);
+      qsize = qsize/sizeof(cl_queue_family_properties_intel);
+      for ( size_t q = 0; q < qsize; q++ ) {
+        if (qfprops[q].capabilities == CL_QUEUE_DEFAULT_CAPABILITIES_INTEL &&
+            qfprops[q].count > partitionMaxSubDevices) {
+          partitionMaxSubDevices = qfprops[q].count;
+        }
+      }
+    } else {
+      return ReturnValue(pi_device_partition_property{0});
+    }
+
+    auto ReturnHelper = [&](auto... Partitions) {
+      struct {
+        pi_device_partition_property Arr[sizeof...(Partitions) + 1];
+      } PartitionProperties = {{Partitions..., 0}};
+      return ReturnValue(PartitionProperties);
+    };
+
+    // Root device
+    if (device->subLevel == -1)
+      device->subLevel = getSubLevel(device);
+    if (device->isRootDevice()) {
+      if (partitionMaxSubDevices < 2) {
+        return ReturnValue(pi_device_partition_property{0});
+      }
+      return ReturnHelper(PI_DEVICE_PARTITION_BY_AFFINITY_DOMAIN);
+    } else if (device->isSubDevice()) {
```

Comment:
No need for `else` as the above `if` early-returns.

Reply:
Will change. Thanks

Reply:
Done